File: src/main/java/ru/anseranser/peshki/ai/BotStrategy.java
package ru.anseranser.peshki.ai;

import ru.anseranser.peshki.engine.*;

import java.util.Comparator;
import java.util.List;

import static ru.anseranser.peshki.engine.Pawn.State.BENCH;
import static ru.anseranser.peshki.engine.Pawn.State.HOMER;

public class BotStrategy {

    public Move selectBestMove(Player player, List<Move> availableMoves, GameConfig config) {
        return availableMoves.stream()
                .max(Comparator.comparingInt(m -> scoreMove(player, m, config)))
                .orElse(null);
    }

    private int scoreMove(Player player, Move move, GameConfig config) {
        if (move.pawn() == null) {
            Cell cornerCell = player.getCorner();
            Pawn existingPawn = cornerCell.getPawn();
            if (existingPawn != null && !existingPawn.getPlayer().equals(player)) {
                return 800_000;
            }
            return 100_000;
        }

        Pawn pawn = move.pawn();
        int steps = move.steps();
        Cell target = pawn.findTargetCell(steps, config);
        if (target == null) return -1;

        boolean killsEnemy = pawn.getState() != Pawn.State.HOMER
                && target.getPawn() != null
                && !target.getPawn().getPlayer().equals(player);

        Player enemy = killsEnemy ? target.getPawn().getPlayer() : null;

        boolean entersHome = pawn.getState() == Pawn.State.FIELDER
                && target.getCellType() == Cell.CellType.HOME;

        Cell source = pawn.getCell();
        boolean onEnemyCorner = source.getCellType() == Cell.CellType.CORNER
                && !source.getOwner().equals(player);
        boolean onOwnCorner = source.getCellType() == Cell.CellType.CORNER
                && source.getOwner().equals(player);

        if (isWinningMove(player, pawn, steps, config)) return 1_000_000;
        if (killsEnemy && isEnemyLastFieldPawn(enemy, config)) return 900_000;
        if (killsEnemy) return 800_000;
        if (entersHome) return 700_000;
        if (onEnemyCorner) return 600_000;
        if (onOwnCorner) return 500_000;

        int maxDistance = config.fieldLength();
        int distance = calculateDistanceToHome(player, pawn, config);
        return 400_000 + (maxDistance - distance);
    }

    private boolean isWinningMove(Player player, Pawn pawn, int steps, GameConfig config) {
        if (pawn.getState() == Pawn.State.HOMER) return false;
        Cell target = pawn.findTargetCell(steps, config);
        if (target == null || target != player.getCorner()) return false;
        return player.getPawns().stream()
                .filter(p -> p != pawn && p.getState() == HOMER)
                .count() == config.numberOfPawns() - 1;
    }

    private boolean isEnemyLastFieldPawn(Player enemy, GameConfig config) {
        if (enemy == null) return false;
        long homeCount = enemy.getPawns().stream()
                .filter(p -> p.getState() == HOMER)
                .count();
        long fieldCount = enemy.getPawns().stream()
                .filter(p -> p.getState() != HOMER && p.getState() != BENCH)
                .count();
        return homeCount == config.numberOfPawns() - 1 && fieldCount == 1;
    }

    private int calculateDistanceToHome(Player player, Pawn pawn, GameConfig config) {
        if (pawn.getState() == Pawn.State.HOMER) return 0;
        Cell current = pawn.getCell();
        if (current == player.getCorner()) return config.fieldLength();
        int distance = 0;
        while (current != player.getCorner() && distance < config.fieldLength()) {
            current = current.getNextFieldCell();
            distance++;
        }
        return distance;
    }
}

File: web/src/ai/BotStrategy.ts
import type { GameConfig } from '../engine/GameConfig';
import { CellType } from '../engine/Cell';
import type { Move } from '../engine/Move';
import { Pawn, PawnState } from '../engine/Pawn';
import type { Player } from '../engine/Player';
import { fieldLength } from '../engine/GameConfig';

export class BotStrategy {
  selectBestMove(player: Player, availableMoves: Move[], config: GameConfig): Move | null {
    return availableMoves.reduce<Move | null>((best, move) => {
      const score = this.scoreMove(player, move, config);
      const bestScore = best ? this.scoreMove(player, best, config) : -Infinity;
      return score > bestScore ? move : best;
    }, null);
  }

  private scoreMove(player: Player, move: Move, config: GameConfig): number {
    if (move.pawn === null) {
      const cornerCell = player.corner;
      const existingPawn = cornerCell.pawn;
      if (existingPawn !== null && existingPawn.player !== player) {
        return 800_000;
      }
      return 100_000;
    }

    const pawn = move.pawn;
    const steps = move.steps;
    const target = pawn.findTargetCell(steps, config);
    if (target === null) return -1;

    const killsEnemy = pawn.state !== PawnState.HOMER
      && target.pawn !== null
      && target.pawn.player !== player;

    const enemy = killsEnemy ? target.pawn!.player : null;

    const entersHome = pawn.state === PawnState.FIELDER
      && target.cellType === CellType.HOME;

    const source = pawn.cell!;
    const onEnemyCorner = source.cellType === CellType.CORNER
      && source.owner !== player;
    const onOwnCorner = source.cellType === CellType.CORNER
      && source.owner === player;

    if (this.isWinningMove(player, pawn, steps, config)) return 1_000_000;
    if (killsEnemy && this.isEnemyLastFieldPawn(enemy!, config)) return 900_000;
    if (killsEnemy) return 800_000;
    if (entersHome) return 700_000;
    if (onEnemyCorner) return 600_000;
    if (onOwnCorner) return 500_000;

    const maxDistance = fieldLength(config);
    const distance = this.calculateDistanceToHome(player, pawn, config);
    return 400_000 + (maxDistance - distance);
  }

  private isWinningMove(player: Player, pawn: Pawn, steps: number, config: GameConfig): boolean {
    if (pawn.state === PawnState.HOMER) return false;
    const target = pawn.findTargetCell(steps, config);
    if (!target || target !== player.corner) return false;
    return player.pawns.filter(p => p !== pawn && p.state === PawnState.HOMER).length === config.numberOfPawns - 1;
  }

  private isEnemyLastFieldPawn(enemy: Player, config: GameConfig): boolean {
    if (!enemy) return false;
    const homeCount = enemy.getPawnsByState(PawnState.HOMER).length;
    const fieldCount = enemy.pawns.filter(p => p.state !== PawnState.HOMER && p.state !== PawnState.BENCH).length;
    return homeCount === config.numberOfPawns - 1 && fieldCount === 1;
  }

  private calculateDistanceToHome(player: Player, pawn: Pawn, config: GameConfig): number {
    if (pawn.state === PawnState.HOMER) return 0;
    let current = pawn.cell!;
    if (current === player.corner) return fieldLength(config);
    let distance = 0;
    while (current !== player.corner && distance < fieldLength(config)) {
      current = current.nextFieldCell!;
      distance++;
    }
    return distance;
  }
}
